packages/hardhat/test/Challenge4.ts

```diff
@@ -22,126 +23,376 @@ describe("🚩 Challenge 4: ⚖️ 🪙 DEX", () => {
   let user2: SignerWithAddress;
   let user3: SignerWithAddress;
 
-  before(async () => {
-    [deployer, user2, user3] = await ethers.getSigners();
+  let contractArtifact: string;
+  if (process.env.CONTRACT_ADDRESS) {
+    contractArtifact = `contracts/${process.env.CONTRACT_ADDRESS}.sol:DEX`;
+  } else {
+    contractArtifact = "contracts/DEX.sol:DEX";
+  }
+
+  function getEventValue(txReceipt: TransactionReceipt, eventNumber: number) {
+    const log = txReceipt.logs.find(log => log.address === dexContract.address);
+    const logDescr = log && dexContract.interface.parseLog(log);
+    const args = logDescr?.args;
+    return args && args[eventNumber]; // index of ethAmount in event
+  }
+
+  async function deployNewInstance() {
+    before("Deploying fresh contracts", async function () {
```

Comment:
Why to add `before` here? I think you can just use `await deployNewInstance()` without before.

Also probably you can use `beforeEach(async () => { await deployNewInstance() })` since that code is repeated. But not sure it applicable for every test

Reply:
> Why to add `before` here? I think you can just use `await deployNewInstance()` without before.

Actually, I had same doubt and also tried removing it but the order of execution breaks which causes test to fail...My understanding is  by adding `before` it makes sure that it runs once before the first `it` in `describe` that too in the correct order. 

> But not sure it applicable for every test

Yeah it's used for only some tests 

Btw I copied tests from https://github.com/scaffold-eth/scaffold-eth-challenges/blob/challenge-4-dex/packages/hardhat/test/constantProductTest.js 😅 and just made tweaks for typescripts but I think we can also update/optimize this test by re-factoring them 🙌